package com.undercut.script

import com.undercut.script.event.Event

abstract class StateMachineScript<T : StateMachineScript<T>> : Script() {
    internal var currentState: State<T>

    abstract fun getStartState(): State<T>

    init {
        currentState = getStartState()
    }

    @Suppress("UNCHECKED_CAST")
    override suspend fun loop() {
        try {
            (currentState.checkNextState(this as T))?.let {
                currentState = it
                return
            }
            currentState.loop(this)
        } catch (exception: Throwable) {
            exception.printStackTrace()
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun onEvent(event: Event) {
        currentState.onEvent(this as T, event)
    }
}

abstract class State<T : StateMachineScript<T>>() {
    abstract suspend fun T.checkNext(): State<T>?
    abstract suspend fun T.stateLoop()
    open fun T.onStateEvent(event: Event) { }

    suspend fun checkNextState(script: T): State<T>? = script.checkNext()
    suspend fun loop(script: T) = script.stateLoop()
    fun onEvent(script: T, event: Event) = script.onStateEvent(event)
}